### PR TITLE
biquad filter sketch (DON'T MERGE)

### PR DESCRIPTION
**Pull request by Stapelzeiger**
I suggest to add a biquad filter like this.
With this we can implement lp/bp/hp filters.
It is computationally simple and there are a lot of resources on how to dimension biquad filters.
We could use this filter for the PID derivative filtering.
What do you think?

### CODE REVIEW

**Comment by Stapelzeiger** [comment]
The implementation is after http://en.wikipedia.org/wiki/Digital_biquad_filter#Transposed_direct_forms

**Comment by pierluca** [comment]
I wouldn't provide a function to automatically create filters to be used within our code.
Manual should be the only way to go for me.

Rationale: manually designed filters can't be hacked on a whim. It forces us to properly design them with an appropriate tool, evaluate their response and integrate them only then. They have to be re-designed and re-evaluated if we are to change even slightly the pass/cut-off bands, thus guaranteeing that we're not suddenly using some weird, fucked up filter without being aware of it.
It's an IIR, it can bite if you're not careful :-)

File: biquad.h [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+							 float a1, float a2);
+
+void filt_biquad_init_lp(filt_biquad_t *f, float Fs, float Fc, float Q);
+
+float filt_biquad_apply(filt_biquad_t *f, float in)
+{
+	float out;
+	out = f->b0 * in + d1;
+	// update filter
+	f->d1 = f->b1 * in + f->a1 * out + f->d2;
```
Comment (pierluca) [inline biquad.h]: `+ f->a1`
should be 
`- f->a1`

same for line below

**Comment by Stapelzeiger** [comment]
Even for a "simple" low-pass? The disadvantage with manual filter coefficients is, that the code is harder to understand and modify. (eg. you think that the cutoff frequency for the pid-derivative-filter might be too low, so you want to try a little higher value, without starting matlab and finding new coefficients.)

File: biquad.h (hunk 1 above)
Comment (Stapelzeiger) [inline biquad.h]: you're right, I'll fix it, same for f->a2

**Comment by pierluca** [comment]
For an order 1 low-pass, you don't need (by definition) a biquad filter nor matlab to select the coefficients :-)

I agree that the code will be hard to modify, but we aren't going to have and updated 10'000 filters. Besides, if you're in a tweaking process, it's easy to keep the code and a filter design tool open in parallel. I don't think it'll be harder to understand as long as the full filter parameters are provided in comment. If this helps, I can provide a MATLAB function that outputs the correct, commented C code using your biquad functions ;-)

It's just my opinion, but I strongly believe having potentially unstable (or bad behaved) IIR filters popping up behind our back is a risk not worth taking as it will ultimately result in longer development times (good luck debugging those in a non-trivial system).

**Comment by Stapelzeiger** [comment]
OK, if you say it's dangerous, I'll keep only the manual initializer.

**Comment by pierluca** [comment]
Thinking about it, I'd also add a function checking that the poles are all in the unit circle and using it to assert it in the manual initializer. Just to be safe.

**Comment by antoinealb** [comment]
Do we really want all poles in unit circle ? I know it makes the system unstable, but sometimes that's desired behaviour, no ? Or maybe add an unsafe function that allows it ?

**Comment by Stapelzeiger** [comment]
I think it's a good idea, but we shouldn't use the assert() macro for that, because it's not a programming logic bug.
I'm not sure what the best option for signaling the error is, maybe just an error return code?

**Comment by pierluca** [comment]
> I know it makes the system unstable, but sometimes that's desired behaviour

Why would you ever want something that goes to +/- infinity?

> I think it's a good idea, but we shouldn't use the assert() macro for that, because it's not a programming logic bug.

A compile-time error would be even better :-P
However, you're right. Still, it should fail as early as possible and in the most noticeable manner.
I'm open to anything along those lines.

**Pull request closed without merging**